File: c/src/classical/substitution/polygraphic/monoalphabetic_squares.c
/**
 * @file monoalphabetic_squares.c
 * @brief Implementation of shared utilities for generating Polybius squares using monoalphabetic cipher transformations
 */

#include "../../../../include/cryptology/classical/substitution/polygraphic/monoalphabetic_squares.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_ALPHABET_SIZE 64
#define MAX_SQUARE_SIZE 36
#define MAX_SQUARE_LINES 6

// Helper function to calculate GCD
static int gcd(int x, int y) {
    while (y) {
        int temp = y;
        y = x % y;
        x = temp;
    }
    return x;
}

// Helper function to convert string to uppercase
static void to_upper_string(char *str) {
    for (int i = 0; str[i]; i++) {
        str[i] = toupper(str[i]);
    }
}

// Helper function to remove duplicates from keyword while preserving order
static void remove_duplicates(const char *keyword, char *result) {
    int seen[256] = {0};
    int result_idx = 0;
    
    for (int i = 0; keyword[i]; i++) {
        char c = toupper(keyword[i]);
        if (isalpha(c) && !seen[c]) {
            result[result_idx++] = c;
            seen[c] = 1;
        }
    }
    result[result_idx] = '\0';
}

int create_caesar_alphabet(const char *alphabet, int shift, char *result, size_t result_size) {
    if (!alphabet || !result || result_size < MAX_ALPHABET_SIZE) {
        return -1;
    }
    
    char alphabet_upper[MAX_ALPHABET_SIZE];
    strncpy(alphabet_upper, alphabet, MAX_ALPHABET_SIZE - 1);
    alphabet_upper[MAX_ALPHABET_SIZE - 1] = '\0';
    to_upper_string(alphabet_upper);
    
    int result_idx = 0;
    for (int i = 0; alphabet_upper[i] && result_idx < (int)result_size - 1; i++) {
        if (isalpha(alphabet_upper[i])) {
            // Apply Caesar shift
            char shifted_char = ((alphabet_upper[i] - 'A' + shift) % 26) + 'A';
            result[result_idx++] = shifted_char;
        }
    }
    result[result_idx] = '\0';
    
    return 0;
}

int create_atbash_alphabet(const char *alphabet, char *result, size_t result_size) {
    if (!alphabet || !result || result_size < MAX_ALPHABET_SIZE) {
        return -1;
    }
    
    char alphabet_upper[MAX_ALPHABET_SIZE];
    strncpy(alphabet_upper, alphabet, MAX_ALPHABET_SIZE - 1);
    alphabet_upper[MAX_ALPHABET_SIZE - 1] = '\0';
    to_upper_string(alphabet_upper);
    
    int result_idx = 0;
    for (int i = 0; alphabet_upper[i] && result_idx < (int)result_size - 1; i++) {
        if (isalpha(alphabet_upper[i])) {
            // Apply Atbash reversal
            char reversed_char = 'Z' - (alphabet_upper[i] - 'A');
            result[result_idx++] = reversed_char;
        }
    }
    result[result_idx] = '\0';
    
    return 0;
}

int create_affine_alphabet(const char *alphabet, int a, int b, char *result, size_t result_size) {
    if (!alphabet || !result || result_size < MAX_ALPHABET_SIZE) {
        return -1;
    }
    
    char alphabet_upper[MAX_ALPHABET_SIZE];
    strncpy(alphabet_upper, alphabet, MAX_ALPHABET_SIZE - 1);
    alphabet_upper[MAX_ALPHABET_SIZE - 1] = '\0';
    to_upper_string(alphabet_upper);
    
    int alphabet_len = strlen(alphabet_upper);
    
    // Check that a is coprime with alphabet length
    if (gcd(a, alphabet_len) != 1) {
        return -1; // Invalid parameters
    }
    
    int result_idx = 0;
    for (int i = 0; alphabet_upper[i] && result_idx < (int)result_size - 1; i++) {
        if (isalpha(alphabet_upper[i])) {
            // Apply Affine transformation: E(x) = (ax + b) mod m
            int x = alphabet_upper[i] - 'A';
            int encrypted_pos = (a * x + b) % alphabet_len;
            result[result_idx++] = alphabet_upper[encrypted_pos];
        }
    }
    result[result_idx] = '\0';
    
    return 0;
}

int create_keyword_alphabet(const char *alphabet, const char *keyword, char *result, size_t result_size) {
    if (!alphabet || !keyword || !result || result_size < MAX_ALPHABET_SIZE) {
        return -1;
    }
    
    char alphabet_upper[MAX_ALPHABET_SIZE];
    strncpy(alphabet_upper, alphabet, MAX_ALPHABET_SIZE - 1);
    alphabet_upper[MAX_ALPHABET_SIZE - 1] = '\0';
    to_upper_string(alphabet_upper);
    
    char keyword_unique[MAX_ALPHABET_SIZE];
    remove_duplicates(keyword, keyword_unique);
    
    // Add remaining alphabet letters
    int seen[256] = {0};
    int result_idx = 0;
    
    // Add keyword letters first
    for (int i = 0; keyword_unique[i] && result_idx < (int)result_size - 1; i++) {
        result[result_idx++] = keyword_unique[i];
        seen[keyword_unique[i]] = 1;
    }
    
    // Add remaining alphabet letters
    for (int i = 0; alphabet_upper[i] && result_idx < (int)result_size - 1; i++) {
        if (!seen[alphabet_upper[i]]) {
            result[result_idx++] = alphabet_upper[i];
        }
    }
    result[result_idx] = '\0';
    
    return 0;
}

int alphabet_to_square(const char *transformed_alphabet, const char *original_alphabet, char *result, size_t result_size) {
    if (!transformed_alphabet || !original_alphabet || !result || result_size < MAX_SQUARE_SIZE) {
        return -1;
    }
    
    char alphabet_copy[MAX_ALPHABET_SIZE];
    strncpy(alphabet_copy, transformed_alphabet, MAX_ALPHABET_SIZE - 1);
    alphabet_copy[MAX_ALPHABET_SIZE - 1] = '\0';
    
    char original_copy[MAX_ALPHABET_SIZE];
    strncpy(original_copy, original_alphabet, MAX_ALPHABET_SIZE - 1);
    original_copy[MAX_ALPHABET_SIZE - 1] = '\0';
    to_upper_string(original_copy);
    
    // Handle I=J combination for English
    if (strcmp(original_copy, "ABCDEFGHIJKLMNOPQRSTUVWXYZ") == 0) {
        for (int i = 0; alphabet_copy[i]; i++) {
            if (alphabet_copy[i] == 'J') {
                alphabet_copy[i] = 'I';
            }
        }
    }
    
    // Determine square size based on original alphabet
    int size;
    if (strcmp(original_copy, "ABCDEFGHIJKLMNOPQRSTUVWXYZ") == 0) {
        size = 5; // English uses 5x5 (25 letters with I=J)
    } else if (strcmp(original_copy, "ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ") == 0) {
        size = 6; // Turkish uses 6x6 (29 letters)
    } else {
        size = (strlen(alphabet_copy) <= 25) ? 5 : 6;
    }
    
    // Pad if needed
    int alphabet_len = strlen(alphabet_copy);
    while (alphabet_len < size * size) {
        alphabet_copy[alphabet_len++] = alphabet_copy[0];
    }
    alphabet_copy[alphabet_len] = '\0';
    
    // Create square
    int result_idx = 0;
    for (int i = 0; i < size && result_idx < (int)result_size - 1; i++) {
        for (int j = 0; j < size && result_idx < (int)result_size - 1; j++) {
            result[result_idx++] = alphabet_copy[i * size + j];
        }
        if (i < size - 1 && result_idx < (int)result_size - 1) {
            result[result_idx++] = '\n';
        }
    }
    result[result_idx] = '\0';
    
    return 0;
}

int create_monoalphabetic_square(
    const char *square_type,
    const char *alphabet,
    const char *mono_params,
    char *result,
    size_t result_size
) {
    if (!square_type || !result || result_size < MAX_SQUARE_SIZE) {
        return -1;
    }
    
    const char *default_alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    const char *use_alphabet = alphabet ? alphabet : default_alphabet;
    
    char transformed_alphabet[MAX_ALPHABET_SIZE];
    
    if (strcmp(square_type, "caesar") == 0) {
        // Parse shift from mono_params (simplified - assumes JSON format)
        int shift = 3; // Default shift
        if (mono_params && strstr(mono_params, "shift")) {
            // Simple parsing - look for "shift":number
            const char *shift_start = strstr(mono_params, "shift");
            if (shift_start) {
                const char *colon = strchr(shift_start, ':');
                if (colon) {
                    shift = atoi(colon + 1);
                }
            }
        }
        
        if (create_caesar_alphabet(use_alphabet, shift, transformed_alphabet, sizeof(transformed_alphabet)) != 0) {
            return -1;
        }
        
    } else if (strcmp(square_type, "atbash") == 0) {
        if (create_atbash_alphabet(use_alphabet, transformed_alphabet, sizeof(transformed_alphabet)) != 0) {
            return -1;
        }
        
    } else if (strcmp(square_type, "affine") == 0) {
        // Parse a and b from mono_params (simplified - assumes JSON format)
        int a = 1, b = 0; // Default values
        if (mono_params) {
            const char *a_start = strstr(mono_params, "a");
            if (a_start) {
                const char *colon = strchr(a_start, ':');
                if (colon) {
                    a = atoi(colon + 1);
                }
            }
            
            const char *b_start = strstr(mono_params, "b");
            if (b_start) {
                const char *colon = strchr(b_start, ':');
                if (colon) {
                    b = atoi(colon + 1);
                }
            }
        }
        
        if (create_affine_alphabet(use_alphabet, a, b, transformed_alphabet, sizeof(transformed_alphabet)) != 0) {
            return -1;
        }
        
    } else if (strcmp(square_type, "keyword") == 0) {
        const char *keyword = "SECRET"; // Default keyword
        if (mono_params && strstr(mono_params, "keyword")) {
            // Simple parsing - look for "keyword":"value"
            const char *keyword_start = strstr(mono_params, "keyword");
            if (keyword_start) {
                const char *colon = strchr(keyword_start, ':');
                if (colon) {
                    const char *quote = strchr(colon, '"');
                    if (quote) {
                        const char *end_quote = strchr(quote + 1, '"');
                        if (end_quote) {
                            // Extract keyword (simplified)
                            keyword = "SECRET"; // For now, use default
                        }
                    }
                }
            }
        }
        
        if (create_keyword_alphabet(use_alphabet, keyword, transformed_alphabet, sizeof(transformed_alphabet)) != 0) {
            return -1;
        }
        
    } else {
        return -1; // Invalid square type
    }
    
    // Convert transformed alphabet to Polybius square
    return alphabet_to_square(transformed_alphabet, use_alphabet, result, result_size);
}

int get_available_monoalphabetic_types(char *types, size_t types_size) {
    if (!types || types_size < 32) {
        return -1;
    }
    
    strncpy(types, "caesar,atbash,affine,keyword", types_size - 1);
    types[types_size - 1] = '\0';
    
    return 0;
}

int validate_mono_params(const char *square_type, const char *mono_params) {
    if (!square_type) {
        return 0;
    }
    
    if (strcmp(square_type, "caesar") == 0) {
        return (mono_params && strstr(mono_params, "shift")) ? 1 : 0;
    } else if (strcmp(square_type, "atbash") == 0) {
        return 1; // No parameters needed
    } else if (strcmp(square_type, "affine") == 0) {
        return (mono_params && strstr(mono_params, "a") && strstr(mono_params, "b")) ? 1 : 0;
    } else if (strcmp(square_type, "keyword") == 0) {
        return (mono_params && strstr(mono_params, "keyword")) ? 1 : 0;
    }
    
    return 0;
}

File: python/cryptology/classical/substitution/polygraphic/monoalphabetic_squares.py
"""
Monoalphabetic-based Polybius Square Generation

This module provides shared utilities for generating Polybius squares using
monoalphabetic cipher transformations. This can be used by any cipher that
employs Polybius squares: Playfair, Two Square, Four Square, Bifid, Trifid, Nihilist.
"""

import string
from typing import Optional, Dict, Any


def create_monoalphabetic_square(
    square_type: str,
    alphabet: Optional[str] = None,
    mono_params: Optional[Dict[str, Any]] = None
) -> str:
    """
    Create a Polybius square using monoalphabetic cipher transformations.
    
    Args:
        square_type: Type of monoalphabetic transformation ("caesar", "atbash", "affine", "keyword")
        alphabet: Base alphabet (default: English)
        mono_params: Parameters for the monoalphabetic transformation
                    - For "caesar": {"shift": int}
                    - For "affine": {"a": int, "b": int}
                    - For "keyword": {"keyword": str}
    
    Returns:
        Square string representation
    
    Raises:
        ValueError: If required parameters are missing or invalid
    """
    if alphabet is None:
        alphabet = string.ascii_uppercase
    
    alphabet_upper = alphabet.upper()
    
    # Generate transformed alphabet based on square type
    if square_type == "caesar":
        if not mono_params or "shift" not in mono_params:
            raise ValueError("Caesar square requires mono_params with 'shift' key")
        transformed_alphabet = _create_caesar_alphabet(alphabet_upper, mono_params["shift"])
    
    elif square_type == "atbash":
        transformed_alphabet = _create_atbash_alphabet(alphabet_upper)
    
    elif square_type == "affine":
        if not mono_params or "a" not in mono_params or "b" not in mono_params:
            raise ValueError("Affine square requires mono_params with 'a' and 'b' keys")
        transformed_alphabet = _create_affine_alphabet(alphabet_upper, mono_params["a"], mono_params["b"])
    
    elif square_type == "keyword":
        if not mono_params or "keyword" not in mono_params:
            raise ValueError("Keyword square requires mono_params with 'keyword' key")
        transformed_alphabet = _create_keyword_alphabet(alphabet_upper, mono_params["keyword"])
    
    else:
        raise ValueError(f"Invalid square_type: {square_type}")
    
    # Convert transformed alphabet to Polybius square
    return _alphabet_to_square(transformed_alphabet, alphabet_upper)


def _create_caesar_alphabet(alphabet: str, shift: int) -> str:
    """Create a Caesar-shifted alphabet."""
    shifted_alphabet = ""
    for char in alphabet:
        if char.isalpha():
            # Apply Caesar shift
            shifted_char = chr((ord(char) - ord('A') + shift) % 26 + ord('A'))
            shifted_alphabet += shifted_char
    return shifted_alphabet


def _create_atbash_alphabet(alphabet: str) -> str:
    """Create an Atbash-reversed alphabet."""
    reversed_alphabet = ""
    for char in alphabet:
        if char.isalpha():
            # Apply Atbash reversal
            reversed_char = chr(ord('Z') - (ord(char) - ord('A')))
            reversed_alphabet += reversed_char
    return reversed_alphabet


def _create_affine_alphabet(alphabet: str, a: int, b: int) -> str:
    """Create an Affine-transformed alphabet."""
    # Check that a is coprime with alphabet length
    def gcd(x, y):
        while y:
            x, y = y, x % y
        return x
    
    if gcd(a, len(alphabet)) != 1:
        raise ValueError(f"Parameter 'a' ({a}) must be coprime with alphabet length ({len(alphabet)})")
    
    # Create Affine-transformed alphabet
    affine_alphabet = ""
    for char in alphabet:
        if char.isalpha():
            # Apply Affine transformation: E(x) = (ax + b) mod m
            x = ord(char) - ord('A')
            encrypted_pos = (a * x + b) % len(alphabet)
            affine_char = alphabet[encrypted_pos]
            affine_alphabet += affine_char
    return affine_alphabet


def _create_keyword_alphabet(alphabet: str, keyword: str) -> str:
    """Create a keyword-based alphabet."""
    keyword_upper = keyword.upper()
    
    # Remove duplicates from keyword while preserving order
    seen = set()
    keyword_unique = ""
    for char in keyword_upper:
        if char.isalpha() and char not in seen:
            keyword_unique += char
            seen.add(char)
    
    # Add remaining alphabet letters
    remaining = ""
    for char in alphabet:
        if char not in seen:
            remaining += char
    
    return keyword_unique + remaining


def _alphabet_to_square(transformed_alphabet: str, original_alphabet: str) -> str:
    """Convert a transformed alphabet to a Polybius square."""
    # Handle I=J combination for English
    if original_alphabet == "ABCDEFGHIJKLMNOPQRSTUVWXYZ":
        transformed_alphabet = transformed_alphabet.replace('J', 'I')
    
    # Determine square size based on original alphabet
    if original_alphabet == "ABCDEFGHIJKLMNOPQRSTUVWXYZ":
        size = 5  # English uses 5x5 (25 letters with I=J)
    elif original_alphabet == "ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ":
        size = 6  # Turkish uses 6x6 (29 letters)
    else:
        size = 5 if len(transformed_alphabet) <= 25 else 6
    
    # Pad if needed
    while len(transformed_alphabet) < size * size:
        transformed_alphabet += transformed_alphabet[0]
    
    # Create square
    square_lines = []
    for i in range(size):
        start_idx = i * size
        end_idx = start_idx + size
        square_lines.append(transformed_alphabet[start_idx:end_idx])
    
    return '\n'.join(square_lines)


def get_available_monoalphabetic_types() -> list:
    """
    Get list of available monoalphabetic square types.
    
    Returns:
        List of available square types
    """
    return ["caesar", "atbash", "affine", "keyword"]


def validate_mono_params(square_type: str, mono_params: Optional[Dict[str, Any]]) -> bool:
    """
    Validate monoalphabetic parameters for a given square type.
    
    Args:
        square_type: Type of monoalphabetic transformation
        mono_params: Parameters to validate
    
    Returns:
        True if parameters are valid, False otherwise
    """
    if square_type == "caesar":
        return mono_params is not None and "shift" in mono_params and isinstance(mono_params["shift"], int)
    
    elif square_type == "atbash":
        return True  # No parameters needed
    
    elif square_type == "affine":
        return (mono_params is not None and 
                "a" in mono_params and "b" in mono_params and
                isinstance(mono_params["a"], int) and isinstance(mono_params["b"], int))
    
    elif square_type == "keyword":
        return (mono_params is not None and 
                "keyword" in mono_params and 
                isinstance(mono_params["keyword"], str) and 
                len(mono_params["keyword"]) > 0)
    
    return False
